File: javascript/53.maximum-subarray.js
/*
 * @lc app=leetcode id=53 lang=javascript
 *
 * [53] Maximum Subarray
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  let pre = nums[0],
    result = nums[0];
  for (let i = 1; i < nums.length; i++) {
    pre = Math.max(nums[i], pre + nums[i]);
    result = Math.max(result, pre);
  }
  return result;
};
// @lc code=end

File: rust/53.maximum-subarray.rs
/*
 * @lc app=leetcode id=53 lang=rust
 *
 * [53] Maximum Subarray
 */

// @lc code=start
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let (mut p, mut result) = (nums[0], nums[0]);
        for i in 1..nums.len() {
            p = nums[i].max(p + nums[i]);
            result = result.max(p);
        }
        result
    }
}
// @lc code=end
